soroban-sdk/src/env.rs

```diff
@@ -447,6 +447,15 @@ impl Env {
     pub fn logs(&self) -> Logs {
         Logs::new(self)
     }
+
+    pub(crate) fn check_same_env(&self, _other: &Self) {        
+        #[cfg(not(target_family = "wasm"))]
+        self.env_impl
+            .check_same_env(&_other.env_impl)
+            .unwrap_optimized()
+        // For Wasm builds there is only one (guest) env, so this check is a
+        // no-op.
```

Comment:
`is_same_env` calls `check_same_env` and needs to be able to handle the fallible case. So this function needs to return a bool or Result like it did before.

https://github.com/stellar/rs-soroban-sdk/blob/f46e9e0610213bbb72285566f9dd960ff96d03d8/soroban-sdk/src/env.rs#L452-L458

Reply:
> is_same_env check has never worked correctly

Given this is a bug in v22, lets try and fix it before merging the bump. I'll take a look on my Monday (Sunday SF time) and see if I can clean up / fix the issues and oddities with `is_same_env` and `check_same_env`.

Reply:
> is_same_env check has never worked correctly

Looks like it's working fine to me.

`Env` `is_same_env` only exists for non-wasm builds, and it calls `Host` `check_same_env` that returns a `Result`. 

So my concern in my initial comment was incorrectly based on thinking `Env` `is_same_env` called `Env` `check_same_env` which isn't the case.

It's a bit confusing that there are `check_same_env` fns named the same that behave differently at different layers, in one returning a Result and in another panicking. So I agree, we should rename the `Host` `check_same_env -> Result` to `Host` `is_same_env -> bool`. Opened issue for that to happen, which is small change:
- https://github.com/stellar/rs-soroban-env/issues/1554

Reply:
Thanks for the update, I've adopted this change.